src/ParseCallMap.hs

```diff
@@ -0,0 +1,313 @@
+import qualified Language.C.Data.Node as CNode
+import qualified Language.C.Data.Position as CPos
+
+-- Grammar of graph files (assumes UTF8 encoding):
+--
+-- The file consists of S-expressions (where the head of each non-empty
+-- subexpression must be an identifier) given by the following grammar,
+-- interpreted by the rules below.  Everything is case-sensitive.  Integer
+-- literals that do not fit in a haskell 'Int' type are silently (and
+-- erroneously) wrapped.
+--
+--  <graph>      ::= <ws> <form>*
+--  <form>       ::= "(" <ws> <identifier> <exp>* ")" <ws> | "(" <ws> ")" <ws>
+--  <exp>        ::= <value> | <form>
+--  <value>      ::= <identifier> | <string_lit> | <int_lit>
+--  <ws>         ::= /([\t\n\r ]|;[^\n]*$)*/
+--  <string_lit> ::= /"([^\\"]|\\[\\"]|)*"/ <ws>
+--  <int_lit>    ::= /-?([1-9][0-9]*)|0([oO][0-7]+|[xX][0-9a-fA-F]+)?/ <ws>
+--  <identifier> ::= /[_:isAlpha:][_:isAlphaNum:]*/ <ws>
+--
+-- The graph file consists of a single callmapForm.
+--
+-- callmapForm          ::= ( callmap <functionForm>* )
+-- functionForm         ::= ( function <identForm> <nodeForm> <actionsForm> <calltreeForm> )
+-- actionsForm          ::= ( actions <permissionActionForm>* )
+-- permissionActionForm ::= ( [need|use|grant|revoke|deny|waive] <identifier> )
+-- calltreeForm         ::= ( calltree <subcallForm>* )
+-- subcallForm          ::= ( call <identForm> )
+--                       |  ( choice <calltreeForm> <calltreeForm> )
+--                       |  ( seq <subcallForm>* )
+-- identForm            ::= ( ident <string_lit> <nodeForm> )
+-- nodeForm             ::= noNode
+--                       |  ( node <cspanForm> )
+-- cspanForm            ::= noPos
+--                       |  builtin
+--                       |  internal
+--                       |  ( span <sourceForm> <int_lit> <sourceForm> )
+-- sourceForm           ::= ( source <string_lit> <int_lit> <int_lit> <int_lit> )
+--
+
+
+-- | An identifier
+newtype Sidentifier = Sidentifier { sidentText :: Text.Text }
+  deriving (Show, Eq, IsString)
+
+-- | Parse the given callmap graph file.  Assumes UTF8 encoding (will use
+-- U+FFFD replacement character on UTF8 decoding errors).
+fromFile :: FilePath -> IO (Either CallMapParseError CallMap)
+fromFile path = do
+  bs <- BSL.readFile path
+  let txt = TextL.decodeUtf8With TextL.lenientDecode bs
+  return (fromSource path txt)
+
+
+-- | @fromSource path txt@ parses @txt@ and returns the resulting 'CallMap'.
+-- (The @path@ is used in error messages)
+fromSource :: FilePath -> TextL.Text -> Either CallMapParseError CallMap
+fromSource = parse parser
+
+parser :: Parser CallMap
+parser = between ws eof callmapForm
+
+callmapForm :: Parser CallMap
+callmapForm = theForm "callmap" (Map.fromList <$> many functionForm)
+
+type CallMapItem = (NodeInfo, CallTree Ident, PermissionActionSet)
+
+functionForm :: Parser (Ident, CallMapItem)
+functionForm = theForm "function" ((,)
+                                    <$> identForm
```

Comment:
Just so we're on the same page - the fact that the alignment of the lines below depends on the length of everything to the left of `(,)` is the issue, yea?  ok.